src/net/atlas/onchain.rs

```diff
@@ -0,0 +1,235 @@
+use util::hash::Hash160;
+use util::hash::MerkleHashFunc;
+use vm::{
+    clarity::{ClarityConnection, ClarityReadOnlyConnection},
+    costs::{ExecutionCost, LimitedCostTracker},
+    database::{ClarityDatabase, ClaritySerializable, MarfedKV, STXBalance},
+    types::{
+        PrincipalData, QualifiedContractIdentifier, SequenceData, StandardPrincipalData, TupleData,
+    },
+    ClarityName, ContractName, Value,
+};
+
+use std::collections::HashMap;
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct OnchainAttachmentsInventory {
+    pub tip: StacksBlockId,
+    pub info: OnchainAttachmentsInventoryInfo,
+    pub pages_indexes: Vec<u32>,
+    pub pages: HashMap<u32, OnchainAttachmentPage>,
+}
+
+impl OnchainAttachmentsInventory {
+    pub fn empty() -> OnchainAttachmentsInventory {
+        OnchainAttachmentsInventory {
+            tip: StacksBlockId([0x00; 32]),
+            info: OnchainAttachmentsInventoryInfo::empty(),
+            pages_indexes: vec![],
+            pages: HashMap::new(),
+        }
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
+pub struct OnchainAttachmentsInventoryInfo {
+    pub pages_count: u32,
+    pub last_page_len: u32,
+    pub page_size: u32,
+}
+
+impl OnchainAttachmentsInventoryInfo {
+    pub fn empty() -> OnchainAttachmentsInventoryInfo {
+        OnchainAttachmentsInventoryInfo {
+            pages_count: 0,
+            last_page_len: 0,
+            page_size: 0,
+        }
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
+pub struct OnchainAttachmentPage {
+    pub index: u32,
+    pub entries: Vec<Hash160>,
+}
+
+pub struct OnchainInventoryLookup {}
+
+impl OnchainInventoryLookup {
+    pub fn get_attachment_content_hashes_at_page_index(
+        page_index: u32,
+        sortdb: &SortitionDB,
+        chainstate: &mut StacksChainState,
+        tip: &StacksBlockId,
+        options: &ConnectionOptions,
+    ) -> Result<OnchainAttachmentsInventory, net_error> {
+        chainstate.maybe_read_only_clarity_tx(&sortdb.index_conn(), tip, |clarity_tx| {
+            let cost_tracker = LimitedCostTracker::new(options.read_only_call_limit.clone());
+
+            // Get pagination informations
+            let info = OnchainInventoryLookup::get_attachments_inv_info(cost_tracker, clarity_tx)?;
+
+            // Read expected_page
+            let expected_page = clarity_tx.with_clarity_db_readonly(|clarity_db| {
+                OnchainInventoryLookup::get_attachment_content_hashes(page_index, &info, clarity_db)
+            })?;
+
+            let pages_indexes = vec![expected_page.index];
+            let mut pages = HashMap::new();
+            pages.insert(expected_page.index, expected_page);
+
+            Ok(OnchainAttachmentsInventory {
+                tip: tip.clone(),
+                info,
+                pages_indexes,
+                pages,
+            })
+        })
+    }
+
+    pub fn get_attachments_inv_info(
```

Comment:
If you wanted to make the Atlas code more generic, you could have the contract identifier and function live in the Atlas config structure and update this method so it just errors out if the contract's function returns a malformed tuple structure.

Reply:
Yeah I think that a fully generic Atlas is a non goal for 2.0, more a 2.1, there are a few plumbing details to implement.